### PR TITLE
add core functionality

### PR DESCRIPTION
**Pull request by briarsweetbriar**
@Ticketfly/frontenders 

The initial PR for `ember-keyboard`. There are several open issues on this repo for feature requests. If you have an idea for one, please feel free to add.

### CODE REVIEW

File: addon/services/keyboard.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    responderStack.removeObject(responder);
+    responderStack.unshiftObject(responder);
+  },
+
+  deactivate(responder) {
+    this.get('responderStack').removeObject(responder);
+  },
+
+  _initializeListener: on('init', function() {
+    const eventNames = ['keyup', 'keydown'].reduce(function(names, name) {
```
Comment (pzuraq) [inline addon/services/keyboard.js]: This was somewhat confusing to read, perhaps:

``` js
const eventNames = ['keyup', 'keydown'].map((name) => `${name}.ember-keyboard-listener`).join(' '); 
```

File: addon/utils/handle-key-event.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+};
+
+export default function handleKeyEvent(event, responderStack) {
+  let triggeredVariant;
+  const keys = gatherKeys(event);
+  const variants = gatherEventNameVariants(event, keys);
+
+  // Finds the first responder with a registered listener for one of the variants
+  const responder = responderStack.find((responder) => {
+    return triggeredVariant = variants.find((variant) => hasListeners(responder, variant));
```
Comment (pzuraq) [inline addon/utils/handle-key-event.js:46]: It's a shame we have to do it this way :confused: Ideally we would be able to get the names of the responders on the component ahead of time and possibly alphabetize the modifiers or something like that. We should look into it at some point.

**Comment by pzuraq** [comment]
Definitely like the way this is going. Couple questions:
1. How does this affect non-keyboard enabled components? Will they handle key events normally?
2. How do we handle branching key responders? Two or more rendered on the same level at the same time. Should they handle events in parallel?

File: addon/utils/handle-key-event.js (hunk 2 above)
Comment (briarsweetbriar) [inline addon/utils/handle-key-event.js:46]: Yeah, I'm not happy with this either, but can't think of a better approach. I've got another branch we're I'm exploring using web workers to take over this util, but the main challenge is actually doing exactly what you said: getting the names of the responders, or rather, the listeners. We can dive deep into the internals of `__ember_meta__._listeners` to peek at them, but as of 2.0, even what we find there is pretty gnarly. Does not look like a stable api to tie into. If you've got a moment tomorrow, it'd be great to sit with you and go over that stuff. Maybe we can find a way to use it.

**Comment by briarsweetbriar** [comment]
1. Shouldn't affect them at all.
2. Not sure I follow exactly what you're asking, but I think I see what you're getting at. On the one hand, `ember-keyboard` has no concept of 'same level' responders. Since they're ordered in a stack, one will always take precedence over another. That being said, responders can co-exist. So long as a prior responder didn't handle an event, the event will bubble up to responders higher in the stack. This allows for functionality that seems almost parallel, such as a `search-bar` component and `nav-bar` existing side-by-side. If you're interested, there's a more detailed example at the end of the readme.

**Comment by pzuraq** [comment]
What I'm thinking about regarding number two is the situation I encountered with `ui-option-list` where there were multiple options rendered at once. I guess you could manually control activation so that only component is active at once, which makes sense and makes it pretty similar to `key-responder`.

**Comment by pzuraq** [comment]
Looks good for a first pass, going to merge. @null-null-null Where are you planning on field testing this?

**Comment by briarsweetbriar** [comment]
@pzuraq Maybe we could try bringing it into `ui-option-list`.

**Comment by pzuraq** [comment]
Sounds good, we could try it. I can't merge at the moment, can you add me as a collaborator?

**Comment by briarsweetbriar** [comment]
@pzuraq Thanks for reviewing so late!

**Pull request merged**